python-checks/src/main/java/org/sonar/python/checks/SklearnPipelineParameterAreCorrectCheck.java

```diff
@@ -0,0 +1,320 @@
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(CallExpression.class))
+      .map(CallExpression::calleeSymbol)
+      .filter(symbol -> symbol.is(Symbol.Kind.CLASS) && !Objects.equals(symbol.fullyQualifiedName(), "sklearn.pipeline.Pipeline"))
+      .map(ClassSymbol.class::cast);
+
+    return stepName.flatMap(stepName1 -> classifierName.map(classifierName1 -> new Result(stepName1, classifierName1)));
+  }
+
+  private record Result(String stepName, ClassSymbol classifierName) {
+  }
+
+  private static Map<String, Set<StringAndTree>> getStepAndParametersFromArguments(CallExpression callExpression) {
+    return callExpression.arguments().stream().filter(argument -> argument.is(Tree.Kind.REGULAR_ARGUMENT))
+      .map(RegularArgument.class::cast)
+      .map(RegularArgument::keywordArgument)
+      .filter(Objects::nonNull)
+      .map(SklearnPipelineParameterAreCorrectCheck::getStepAndParameterFromName)
+      .<StepAndParameter>mapMulti(Optional::ifPresent)
+      .collect(mergeStringAndTreeToMapCollector());
+  }
+
+  private static Map<String, Set<StringAndTree>> getStepAndParametersFromDict(CallExpression callExpression) {
+    return Optional.ofNullable(TreeUtils.nthArgumentOrKeyword(1, "param_grid", callExpression.arguments()))
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(RegularArgument.class))
+      .map(RegularArgument::expression)
+      .stream()
+      .flatMap(SklearnPipelineParameterAreCorrectCheck::extractKeyValuePairFromDictLiteral)
+      .map(KeyValuePairImpl::key)
+      .map(TreeUtils.toInstanceOfMapper(StringLiteral.class))
+      .filter(Objects::nonNull)
+      .map(stringLiteral -> getStepAndParameterFromString(stringLiteral.trimmedQuotesValue(), stringLiteral))
+      .<StepAndParameter>mapMulti(Optional::ifPresent)
+      .collect(
+        mergeStringAndTreeToMapCollector());
+  }
+
+  private static Stream<KeyValuePairImpl> extractKeyValuePairFromDictLiteral(Expression expression) {
+    return Optional.of(expression)
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(Name.class))
+      .flatMap(Expressions::singleAssignedNonNameValue)
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(DictionaryLiteralImpl.class))
+      .map(DictionaryLiteral::elements)
+      .stream()
+      .flatMap(Collection::stream)
+      .map(TreeUtils.toInstanceOfMapper(KeyValuePairImpl.class))
+      .filter(Objects::nonNull);
+  }
+
+  private static Collector<StepAndParameter, ?, Map<String, Set<StringAndTree>>> mergeStringAndTreeToMapCollector() {
+    return Collectors.toMap(StepAndParameter::step, stepAndParameter -> Set.of(new StringAndTree(stepAndParameter.parameter, stepAndParameter.location)),
+      (set1, set2) -> {
+        var set = new HashSet<>(set1);
+        set.addAll(set2);
+        return set;
+      });
+  }
+
+  private record StringAndTree(String string, Tree tree) {
+  }
+
+  private static Optional<StepAndParameter> getStepAndParameterFromName(Name name) {
```

Comment:
Maybe here the usage of a regex could help in just splitting the last `__`. This could also help in the creation of StepAndParameter as it could be reused for both, the string and the name.